### ISSUE TITLE
filters.hag_dem: Unexpected performance

### ISSUE THREAD

**Issue by xycarto** [PDAL/PDAL#3828]
I am running a hag_dem filter using `pdal 2.4.1 (git-version: eb2616)`

Pipeline JSON
```
{
    "pipeline": [
        {
            "type": "readers.las"
        },
        {
            "type": "filters.hag_dem"
        },
        {
            "type": "writers.las",
            "extra_dims": "HeightAboveGround=float32",
            "minor_version": 4
        }
    ]
}
```

My stats for the Z value of the input LAZ (pdal info --stats):
```
{
        "average": 1901.357866,
        "count": 6305939,
        "maximum": 2239.13,
        "minimum": 1873.54,
        "name": "Z",
        "position": 2,
        "stddev": 31.37241112,
        "variance": 984.2281798
},
```

My stats for the min max of the input raster (gdalinfo -mm):
```
Metadata:
    STATISTICS_MAXIMUM=2225.78
    STATISTICS_MEAN=1887.8240087481
    STATISTICS_MINIMUM=1873.96
    STATISTICS_STDDEV=28.135564876022
    STATISTICS_VALID_PERCENT=76.46
```

If I run the above pipeline, I get these stats for my HeightAboveGround value in my "normalized" LAZ:

```
{
        "average": 5.697506138,
        "count": 6305939,
        "maximum": 12217.59961,
        "minimum": -28.54008675,
        "name": "HeightAboveGround",
        "position": 16,
        "stddev": 25.82151762,
        "variance": 666.7507722
      }
```

Both input projection and extent match for EPSG:3310.

I would expect the HeightAboveGround values to be within a more normal range like min:0 and max:120, given the min/max value ranges of the inputs

I am curious if this might be bug or if this expected behavior in the `filters.hag_dem`?  I can set up data downloads if you are curious to test.

**Comment by xycarto** [comment]
Running `hag_nn` on the input LAZ gives:

```
{
        "average": 5.635608394,
        "count": 6305939,
        "maximum": 343.269989,
        "minimum": -28.94000053,
        "name": "HeightAboveGround",
        "position": 16,
        "stddev": 9.015286739,
        "variance": 81.27539498
}
```

And running `hag_delaunay` produces:

```
,
      {
        "average": 5.635270726,
        "count": 6305939,
        "maximum": 343.269989,
        "minimum": -28.89999962,
        "name": "HeightAboveGround",
        "position": 16,
        "stddev": 9.015599112,
        "variance": 81.28102735
      }
```

**Comment by hobu** [comment]
Some big outliers in there? You'll have to post some data for anyone to actually do an investigation.

**Comment by xycarto** [comment]
Collection of test data: https://drive.google.com/drive/folders/1yXdHni-SjI2uSSvxPkxpj7qtaVh3_MR8?usp=sharing

Thanks for the response.  If it is an outlier issue, I'd be curious why `pdal info --stats` doesn't pick those values up in the statistics. It not really really an issue, but something I could be aware of when developing tests.

**Comment by adamsteer** [comment]
Hi, I took a quick look - and wondered if there's something going on with PDAL trying to get a height where there is a nodata value in the DEM. This train of thought turned up because the combination of unsigned nodata value, elevation range and normalised height add up to about the outlier value (12 000 ish).

Would setting all pixels in the lake to 'lake elevation' instead of nodata help? it might be that nodata values in the water body are being somehow interpreted as real, inside the neighborhood of the point being queried for height above DEM. 

I took a quick look at the code and didn't see an obvious way that hag_dem handles (ignores) DEM nodata values - however, my ability to interpret c++ and how gdal works should be taken as super lightweight, it is very likely I missed something.

**Comment by adamsteer** [comment]
I think my hunch about weird DEM values has some more weight now. After a lot of gdal wrangling to change the `nodata` value in the supplied DEM to -99 (and convince the data to not have -9999 as any kind of value anymore), I ran the `hag_dem` pipeline and got:

`{
        "average": 5.658257423,
        "count": 6305939,
        "maximum": 2317.600098,
        "minimum": -28.54008675,
        "name": "HeightAboveGround",
        "position": 16,
        "stddev": 10.06828974,
        "variance": 101.3704582
      }`

Key point is that `maximum` in `HeightAboveGround` changed to reflect adjusted DEM values.

Doing `hag_dem` and ferrying `HeightAboveGround` to `Z` gives a bit of confirmation - there are only a few 'weird points'.  In this first picture, the original laz file is a blackish blob, blue / red points are HAG, computed using the originally supplied DEM.
![Screenshot from 2022-08-05 16-37-38](https://user-images.githubusercontent.com/13084118/183101388-30b3f4bc-41d0-4ff8-ba43-d365c841f307.png)

In the next picture, same - but with modified `nodata` value - to -99. It was quite a bit of work to convince the DEM that it should not have values of -9999, so I am not 100% convinced about my earlier guess that `hag_dem` isn't handling `nodata` well.
![Screenshot from 2022-08-05 16-45-37](https://user-images.githubusercontent.com/13084118/183101682-16b4e390-641a-4f90-9184-782edde74148.png)

I guess for now take a look at DEM creation? The point of this exercise was to see if `hag_dem` was generating the huge outliers, or if the outliers responded to DEM values. I think it is pretty safe to say that the latter is true.

**Comment by hobu** [comment]
Yes hag_dem is not ignoring nodata at the moment. We should have a patch soon that fixes this for 2.5.0

**Comment by xycarto** [comment]
Thank you all for looking into this.  It is very helpful to confirm the expected behavior. I can easily plan and patch on my end with this knowledge.
